.teamcity/_self/projects/WPComTests.kt

```diff
@@ -219,16 +221,9 @@ fun gutenbergPlaywrightBuildType( targetDevice: String, buildUuid: String ): Bui
 		name = "Playwright E2E Tests ($targetDevice)"
 		description = "Runs Gutenberg E2E tests as $targetDevice using Playwright"
 
-		artifactRules = """
-			logs.tgz => logs.tgz
-			screenshots => screenshots
-			trace => trace
-		""".trimIndent()
+		artifactRules = artifactRules()
```

Comment:
What if we create an extension of `BuildType` (for example, `e2eBuildType`) that already has common parts (vcs, artifacts, features) already assigned? It doesn't feel very OOP-y to create an object `BuildType` and always assign the same value to the same parameters.

Reply:
That's definitely an idea...I'll take a stab at it. With the time we have left counting down I want to make sure I get this done before you leave.

Reply:
I created a custom build type as follows:

```
fun calypsoE2EBuildType(): BuildType {
  // build type definitions
  steps {
      prepareEnvironment()
      collectResults()
  }
}
```

Note the partially-populated `steps`. I then want to call this from a build configuration:

```
var config = calypsoE2EBuildType()

config.steps.items.add(1,  runTests( <params> ) )
```

In the above, I want to insert the `runTests` step in between the `prepareEnvironment()` and `collectResults` steps, but this does not appear to work.

Reply:
Never mind, I used the `WPComPluginBuild` class as reference - my mistake was trying to use a `fun` for that purpose.

Now I am blocked by an odd issue where if I use the following combination:

Playwright E2E build configuration
-> calls runTests
    -> calls bashNodeScript

the dependencies are not being recognized as having been installed on the system, despite it being so:

```
   browserType.launch: Host system is missing dependencies!

11:59:36   

11:59:36     Missing libraries are:

11:59:36         libnss3.so

11:59:36         libnssutil3.so

11:59:36         libsmime3.so

11:59:36         libnspr4.so

11:59:36         libatk-1.0.so.0

11:59:36         libatk-bridge-2.0.so.0

11:59:36         libcups.so.2

11:59:36         libdrm.so.2

11:59:36         libdbus-1.so.3

11:59:36         libxkbcommon.so.0

11:59:36         libXcomposite.so.1

11:59:36         libXdamage.so.1

11:59:36         libXfixes.so.3

11:59:36         libXrandr.so.2

11:59:36         libgbm.so.1

11:59:36         libasound.so.2

11:59:36         libatspi.so.0
```

(build https://teamcity.a8c.com/buildConfiguration/calypso_Calypso_E2E_Playwright_mobile/7175344?buildTab=log&focusLine=484&expandAll=true)

This issue goes away if I change the call structure to:

Playwright E2E build configuration
-> calls bashNodeScript

The build configuration is able to execute the tests as expected.

(build https://teamcity.a8c.com/buildConfiguration/calypso_Calypso_E2E_Playwright_mobile/7175410)

Rather confusing as to why this is the case, but I'm still investigating.

Reply:
I think I've got it.

It was due to the missing line `dockerImage = "%docker_image_e2e%"`.

Reply:
Resolved. Now using `class E2EBuildType` which sets these values as default.